x-pack/filebeat/input/awss3/sqs_acker.go

```diff
@@ -0,0 +1,185 @@
+	// We want to execute the logic of this call only once, when the ack mutex was locked on init
+	if !a.ackMutexLockedOnInit.Load() {
+		return
+	}
+
+	a.isSQSAcker = true
+
+	a.msg = msg
+	a.EventsToBeAcked = atomic.NewUint64(publishedEvent)
+	a.ReceiveCount = receiveCount
+	a.start = start
+	a.processingErr = processingErr
+	a.Handles = handles
+	a.keepaliveCancel = keepaliveCancel
+	a.keepaliveWg = keepaliveWg
+	a.msgHandler = msgHandler
+	a.log = log
+	a.client = client
+
+	a.ackMutex.Unlock()
+	a.ackMutexLockedOnInit.Store(false)
+}
+
+func (a *EventACKTracker) FullyAcked() bool {
+	return a.TotalEventsAcked.Load() == a.EventsToBeAcked.Load()
+
+}
+
+// WaitForS3 must be called after SyncEventsToBeAcked
+func (a *EventACKTracker) WaitForS3() {
+	// If it's fully acked then cancel the context.
+	if a.FullyAcked() {
+		a.cancel()
+	}
+
+	// Wait.
+	<-a.ctx.Done()
+}
+
+// FlushForSQS delete related SQS message
+func (a *EventACKTracker) FlushForSQS() {
+	// This is moved here because clientOnlyACKer.ClientClosed() reset the acker listener (libbeat/common/acker/acker.go:333)
+	defer a.client.Close()
+
+	if !a.isSQSAcker {
+		return
+	}
+
+	// Stop keepalive visibility routine before deleting.
+	a.keepaliveCancel()
+	a.keepaliveWg.Wait()
+
+	err := a.msgHandler.DeleteSQS(a.msg, a.ReceiveCount, a.processingErr, a.Handles)
+	a.DeletionWg.Done()
+
+	if err != nil {
+		a.log.Warnw("Failed deleting SQS message.",
+			"error", err,
+			"message_id", *a.msg.MessageId,
+			"elapsed_time_ns", time.Since(a.start))
+	} else {
+		a.log.Debugw("Success deleting SQS message.",
+			"message_id", *a.msg.MessageId,
+			"elapsed_time_ns", time.Since(a.start))
+	}
+}
+
+// ACK decrements the number of total Events ACKed.
+func (a *EventACKTracker) ACK() {
+	// We need to lock on ack mutex, in order to know that we have passed the info to the acker about the total to be acked
+	// But we want to do it only before the info have been passed, once they did, no need anymore to lock on the ack mutext
+	if a.ackMutexLockedOnInit.Load() {
+		a.ackMutex.Lock()
+		defer a.ackMutex.Unlock()
+	}
+
+	if a.FullyAcked() {
+		panic("misuse detected: ACK call on fully acked")
+	}
+
+	a.TotalEventsAcked.Inc()
+
+	if a.FullyAcked() {
+		a.cancel()
+	}
+}
+
+// NewEventACKHandler returns a beat ACKer that can receive callbacks when
+// an event has been ACKed an output. If the event contains a private metadata
+// pointing to an eventACKTracker then it will invoke the trackers ACK() method
+// to decrement the number of pending ACKs.
+func NewEventACKHandler() beat.EventListener {
+	return acker.ConnectionOnly(
+		acker.EventPrivateReporter(func(_ int, privates []interface{}) {
+			for _, current := range privates {
+				if ackTracker, ok := current.(*EventACKTracker); ok {
+					ackTracker.ACK()
+
+					if ackTracker.FullyAcked() {
+						ackTracker.FlushForSQS()
```

Comment:
If this fails you can't retry it. I assume you'll end up reprocessing the message eventually leading to duplicated data ingestion?

Why not have this send to a dedicated deletion goroutine that can handle retrying?

Reply:
To take this to the extreme we could start persisting acknowledged but not yet deleted objects to disk to avoid reprocessing them if the beat restarts. That is definitely a separate enhancement if we ever find out that reprocessing is something we want to avoid when we can.

Reply:
> If this fails you can't retry it. I assume you'll end up reprocessing the message eventually leading to duplicated data ingestion?
> 
> Why not have this send to a dedicated deletion goroutine that can handle retrying?

in the end we will have as much goroutines as before the changes :)

we need to ensure at-least-once delivery and not exactly-once delivery. I won't go so much further to the second

Reply:
I don't know that you need a goroutine per message. You could have a single goroutine reading from a channel. I assume calling `DeleteMessage` is much cheaper than processing an entire s3 object, so it should complete faster.

If you used a buffered channel that could give you a nice mechanism for deleting messages in bulk, since that is possible.

Minimizing the chance of reprocessing a potentially large S3 object pointless has value, but I will agree it isn't the primary point of this PR.

Reply:
At least for the memory queue, I believe these acknowledgement callbacks are already in their own goroutine, you would just have to batch them manually if you wanted to do that later.

https://github.com/elastic/beats/blob/07e231bd3842957119630d4b609b7653c34f5223/libbeat/publisher/queue/memqueue/broker.go#L172-L175

https://github.com/elastic/beats/blob/07e231bd3842957119630d4b609b7653c34f5223/libbeat/publisher/queue/memqueue/ackloop.go#L145-L148

I wanted to make sure the latency of the `DeleteMessages` call here wouldn't block any other part of the pipeline. To me it looks like it only blocks processing the next ack, but there are a lot of layers to this so there's a chance I missed something.

Reply:
> I don't know that you need a goroutine per message. You could have a single goroutine reading from a channel. I assume calling DeleteMessage is much cheaper than processing an entire s3 object, so it should complete faster.
> If you used a buffered channel that could give you a nice mechanism for deleting messages in bulk, since that is possible.

Sorry, I understood you were proposing a dedicated goroutine for each message. Yes, with a single goroutine and a buffered channel we could have also batched deletion

> I wanted to make sure the latency of the DeleteMessages call here wouldn't block any other part of the pipeline.
> To me it looks like it only blocks processing the next ack, but there are a lot of layers to this so there's a chance I missed something.

It seems to me that it blocks processing the next batch of acks:
https://github.com/elastic/beats/blob/ecfb88fa04b544cf30cd83d122941e054c2cc459/libbeat/publisher/queue/memqueue/ackloop.go#L39-L56

I'm not sure if a batch contains acks from different ackers. The first ack of every acker (ie: sqs message)) would be blocked anyway because before we ack the first event we need to wait for all events to be published, so `DeleteMessages` won't block it.



> Minimizing the chance of reprocessing a potentially large S3 object pointless has value, but I will agree it isn't the primary point of this PR.

Oki, if you are fine let's keep optimisation on reprocessing a potentially large S3 object for a separated PR. And also processing batches of deletion. This is something that already miss in the current implementation.

Reply:
Yes we can deal with minimizing unnecessary reprocessing separately.

Making sure the way acks happen is as efficient as possible I would say is in scope here.

> I'm not sure if a batch contains acks from different ackers. 

It could yes, since it is receiving a batch from the output which could contain events published by multiple different calls to `client.Publish` calls.

@faec is probably the best person to comment on exactly how the memory queue ack processing works, but it seems to me if we leave the `FlushForSQS` call where it is then we are always limited to deleting SQS messages sequentially and if there are other events in the queue they can't be acked until any pending DeleteMessage calls complete.

It definitely feels like that could be improved. Doing batched deletion in a separate goroutine seems like a natural improved, especially since we know Vector is also doing batched deletion as a point of comparison.

Reply:
yes, I've observed that in the client event listener the event comes for different ackers.

@faec suppose there are 10 clients and 100 message: each message is processed in a separated goroutine, where the calls to `client.Publish()` are made sequentially. all the events coming from the same sqs message will share the same acker.
each client will handle 10 message, each of them with their own acker.
every single acker waits on its own mutex unlock (once all events are published) before acking the first event

what will be the impact on the event listener? are the dispatch to the listener sequential/must one dispatch to the listener wait for the previous own returning before another dispatch (it does not seems like that)?

if so I expect each acker of a client blocking the following ones, because of its own mutex waiting.

if not, each acker wait only for the first event to be acked on the its mutex

Reply:
@aspacca I'm not sure I understand what that last scenario is getting at with respect to the early parts of the thread... the effect of a mutex would vary depending on how long it is held and what it blocks, and I'm not sure I understand the model you're describing well enough.

However, I can comment on the earlier discussion:
- ACKs are delivered strictly in order. Later ACK callbacks cannot happen until earlier ones are finished.
- Batches can contain a mix of events from any number of producers. There is no correlation between producers and batches (so slowdown in one producer's ACK handling will affect acknowledgments for all producers).
- You're right, delays in an ACK callback will block additional ACKs from being delivered to the queue. Aside from the ACK callbacks themselves, the main issue with this is that it prevents unused space from being recovered, preventing new events from being inserted in the queue, which can potentially starve the output workers.

In general the work done in an ACK callback should be as short as possible. If there's a chance it might block for any length of time, it should be delegated to a helper goroutine rather than done synchronously in the callback. @cmacknz's suggestion of a single deletion goroutine that processes pending deletion requests sounds reasonable to me for the current case.

Reply:
thanks @faec 

> In general the work done in an ACK callback should be as short as possible. If there's a chance it might block for any length of time, it should be delegated to a helper goroutine rather than done synchronously in the callback. @cmacknz's suggestion of a single deletion goroutine that processes pending deletion requests sounds reasonable to me for the current case.

main waiting time for ACK block is not related to the single deletion, but for waiting that all the events for a specific SQS message are fully published. because we need to know the total events to be acked in order to decrease the counter and once it reaches 0 the message is fully acked and we can delete it.

once we have the information of the total amount of events in an SQS message there's no more waiting time for each ACK, but for the final one for message deletion.

> * Batches can contain a mix of events from any number of producers. There is no correlation between producers and batches (so slowdown in one producer's ACK handling will affect acknowledgments for all producers).
> * You're right, delays in an ACK callback will block additional ACKs from being delivered to the queue. Aside from the ACK callbacks themselves, the main issue with this is that it prevents unused space from being recovered, preventing new events from being inserted in the queue, which can potentially starve the output workers.

As the deletation call is not the main slower, maybe we can start a goroutine for each batch, producers inside that goroutine will wait each order according to the order in the batch, when they'll have to wait, but next batches won't wait. do you see any caveat on that?

Reply:
> main waiting time for ACK block is not related to the single deletion, but for waiting that all the events for a specific SQS message are fully published

This part I'm not sure I follow -- if it's just a matter of updating counters, then there should be no need to block in the acknowledgment callbacks to begin with, right? The acknowledgment callbacks should never be waiting for something else to happen, just updating metadata. If they reach the final event of a completed SQS message, they can trigger deletion, otherwise there's no work to do. If the final event of an SQS message isn't yet known, then the SQS message can't be fully acknowledged yet anyway and deletion checking can be deferred til the next call.

> maybe we can start a goroutine for each batch, producers inside that goroutine will wait each order according to the order in the batch, when they'll have to wait, but next batches won't wait. do you see any caveat on that?

Possibly there's some miscommunication here -- when I say "batch," I mean a buffer of sequential events that is read from the queue in one chunk and forwarded to an output worker. The inputs and the producers have no visibility into batch processing, so there's no way for them to know which events are in the same batch to begin with. Input handling should not try to infer anything about batch sequence. All you should need to know is that the events published through a particular client are always acknowledged through that client in strictly the same order. When you publish the last event of an SQS message, you should make note of its index in the event sequence for its beats client, and then free it when the acknowledgments reach that index.

Possibly the difficulty is around keeping track of event states and acking order? [Here is a helper object](https://github.com/elastic/beats/blob/6d8882b5d653176f2355f317754e6967773fabcf/libbeat/common/acker/acker.go#L67) that you can provide to the `EventListener` parameter when creating a pipeline client, to aggregate errors and acknowledgments and report them to the input in published order. (This helper is especially useful because it handles things like events that were filtered out by processors and therefore will never be acknowledged, still reporting them in published order). You can see an example of this helper being used in an input [here](https://github.com/elastic/beats/blob/6d8882b5d653176f2355f317754e6967773fabcf/x-pack/filebeat/input/shipper/input.go#L170) and [here](https://github.com/elastic/beats/blob/6d8882b5d653176f2355f317754e6967773fabcf/x-pack/filebeat/input/shipper/acker.go#L24).

Reply:
> > main waiting time for ACK block is not related to the single deletion, but for waiting that all the events for a specific SQS message are fully published
> This part I'm not sure I follow -- if it's just a matter of updating counters, then there should be no need to block in the acknowledgment callbacks to begin with, right? 

It's not only updating a counter, but knowing when the updated counter is equal to a amount indicating the final event.

> If the final event of an SQS message isn't yet known, then the SQS message can't be fully acknowledged yet anyway and deletion checking can be deferred til the next call.

the final event could be know after every ack is already done, so where will be no next call.

the timeline could be the following:
. T1 client publishes event 1
. T2 event listener acks event 1 (and if the sqs message has only 1 event it should delete the message)
. T3 sqs event processor mark sqs message as processed, informing the acker that there is only 1 event

That's why the blocking so that T2 can happens after T3


I've modified the acker to add a goroutine for checking "if the sqs message has only 1 event it should delete the message".

So now we can. have the timeline, and we don't need to block T2 after T3, because the sqs deletion happens in a separated goroutine.

Indeed `TrackingCounter` is the correct listener to use.

Reply:
> the final event could be know after every ack is already done, so where will be no next call.

Even if the sequencing requires this uncertainty, this could be resolved by checking the latest acked index when confirming the final event of a message, there is no need for the ack callback itself to wait for this.

> but I didn't use in the end, and instead relied on [AddEvent]

`AddEvent` is very much the wrong thing to use here. `AddEvent` indicates that an event is about to enter the pipeline, it has not yet been sent to an output or acknowledged, so using this as confirmation of acknowledgment will definitely cause data loss.